Truncate displayed ship names to prevent overflowing containers

tehhowch opened the pull request:
Adds front or end string truncation to the following locations:
 - Shipyard sale confirmation dialog: each ship occupies a maximum of one line instead of possibly overflowing onto multiple (#2249)
 - Shipyard/Outfitter ship details panel: no overflow out of rounded square border
 - Hailing screen: no overflow off right edge for combinations of long ship names and/or long government names.
 - Targeting reticle: no overflow off left edge of ship name
 - Player Info fleet list: no overflow of ship name across category
 - Ship detail screen: no overlap of "ship:" and the ship name

I have applied front truncation where sensible to preserve any identifying numbers that may have been added to the end of the ship name when buying in bulk.

<details>
<summary>Appearance with this PR:</summary>


Selling multiple ships:
![pr_sell_5](https://user-images.githubusercontent.com/20871346/27143383-ccf06ba4-50f3-11e7-8024-a2c7c909ef04.png)

Shipyard/Outfitter details panel:
![pr_shoppanel](https://user-images.githubusercontent.com/20871346/27144359-1cd3a1b0-50f7-11e7-91ba-5cfdf0000009.png)


Hailing a very long-named ship:
![pr_hail](https://user-images.githubusercontent.com/20871346/27143438-f60f0a22-50f3-11e7-88c5-608a93ceac87.png)

Targeting ships (this is a rather small field due to its position and because it is center-aligned):
![pr_target_1](https://user-images.githubusercontent.com/20871346/27143843-5806e87a-50f5-11e7-8fde-fde2615f1bf0.png)
![pr_target_2](https://user-images.githubusercontent.com/20871346/27143851-5ba382ea-50f5-11e7-858d-de80dfce2a22.png)

Player Info screen:
![pr_playerinfo](https://user-images.githubusercontent.com/20871346/27143611-90c95ffe-50f4-11e7-8b94-e36a47bc0e99.png)

Ship detail screen:
![pr_shipinfo](https://user-images.githubusercontent.com/20871346/27143603-88abe0b2-50f4-11e7-8f1b-d552e28d5cc1.png)
</details>

endless-sky commented on source/ShipyardPanel.cpp:
```diff
@@ -255,16 +257,17 @@ void ShipyardPanel::Sell()
 		auto it = playerShips.begin();
-		for(int i = 0; i < MAX_LIST - 1; ++i)
-			message += (*it++)->Name() + ",\n";
+		message += (*it++)->Name() + ",\n";
```
Is it intentional to not truncate the first ship's name?

tehhowch replied:
It was indeed intentional, because the first line doesn't start with the ship's name and thus has different wrapping behavior than all the other lines.

endless-sky commented:
I'm not sure that it ever makes sense to TruncateFront() instead of just Truncate(). I can't think of any other UI setting where the front of a string is truncated; if the end of the string is significant it's more common for the program to truncate the middle.

tehhowch commented:
@endless-sky Indeed, I agree. But rather than write Font::TruncateMiddle and also make the PR and then have it go days / weeks without comment, I chose to use the already-implemented truncate functions.
Now that I know this PR has a chance of merging, I will happily write Font::TruncateMiddle and use it instead.

tehhowch commented:
* Added function Font::TruncateMiddle using the same interpolation algorithm as the existing Truncate functions.
This function could be tweaked to control the excision location (e.g. prefer whole words / at word breaks), but that introduces a lot of extra code and corner case management for a function that's already not going to see a lot of use.
* Changed all ship-name truncations to be TruncateMiddle except for the hailing panel, as here the non-ship portion of the header is also of varying length. Using TruncateMiddle would result in a different truncation position for the ship than might normally be seen, and could possibly overlap onto the government name if it was long enough as well. Adding an extra line to calculate the remaining width for the ship name after accounting for the width of the government name seemed like too much.
* Adjusted some of the target widths

The shipyard sale dialog is kept in its original format, only truncating ship names when they are forced into a list structure via newlines (since forced newline + long ship name =  #2249 = the reason for this PR). The 1st and last ship in the sale list, except when selling more than 20 ships at once, are not forced into a list structure and can wrap naturally.

<details>
<summary>Appearance</summary>

Selling multiple ships:
![pr_sellmany](https://user-images.githubusercontent.com/20871346/27174822-04a77530-5183-11e7-8e3d-062058433775.png)

Shipyard/Outfitter ship name
![pr_shoppanel](https://user-images.githubusercontent.com/20871346/27174892-34892f0a-5183-11e7-9bbf-ec82974d4ff4.png)

Target Reticle
![pr_targeting](https://user-images.githubusercontent.com/20871346/27174910-41c0cf8e-5183-11e7-9fed-454f24ead526.png)

Player InfoPanel
![pr_playerinfo](https://user-images.githubusercontent.com/20871346/27174927-5090b84e-5183-11e7-928b-8fad4fd1f8d8.png)

Ship InfoPanel
![pr_shipinfo](https://user-images.githubusercontent.com/20871346/27174942-58ba73f2-5183-11e7-863f-c7d933e99542.png)
</details>

Amazinite commented:
Not what it was originally named, but still a good example.
